### PR TITLE
Distributed systems evolution: topology changes

### PR DESCRIPTION
**Pull request by mauroservienti**
## PoA

- [x] Synopsis
- [x] Tags
- [x] Header image
- [x] Header image credits
- [x] Review
- [x] Rename file to match publish date
- [ ] Buffer content

### CODE REVIEW

File: _posts/topology-changes.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ 
+The third option is to introduce a triage endpoint. We will convert Sales into the triage endpoint and add two more endpoints to handle messages: Sales.PostPaymentProcessing and Sales.Operations, for example. The triage endpoint masquerades the processing endpoints topology to senders. Using a forwarding technique, the triage endpoint inspects incoming messages and then redirects them to the processing endpoint.
+
+The second option is designed to be temporary. We will modify senders to properly route messages to the new endpoint and remove the forwarding logic. On the contrary, the third option comes with the goal of not breaking senders.
+
+### Publish/subscribe caveats
+
+All that presented so far applies well to message sending. Pub/sub is a bit trickier and what we do depends more on the underlying infrastructure. Let's say that we're deploying on AWS (Amazon Web Services), and we are using SQS (Simple Queue Service) and SNS (Simple Notification Service).
```
Comment (kbaley) [inline _posts/topology-changes.md:45]:
```suggestion
Everything presented so far applies well to message sending. Pub/sub is a bit trickier and what we do depends more on the underlying infrastructure. Let's say that we're deploying on AWS (Amazon Web Services), and we are using SQS (Simple Queue Service) and SNS (Simple Notification Service).
```

File: _posts/topology-changes.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+Let's imagine that an event is responsible for triggering post-payment operations. Once users pay, the payment processor publishes a `PaymentCompleted` event. Based on that, Salles subscribed to the event and started the post-payment operations. Visually, it's like the following diagram:
```
Comment (kbaley) [inline _posts/topology-changes.md:47]:
```suggestion
Let's imagine that an event is responsible for triggering post-payment operations. Once users pay, the payment processor publishes a `PaymentCompleted` event. Based on that, Sales subscribed to the event and started the post-payment operations. Visually, it's like the following diagram:
```

File: _posts/topology-changes.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    S2 -->Q2[Sales.PostPaymentProcessing - SQS input queue]
+</div>
+
+The Sales endpoint receives the `PaymentCompleted` event and Sales.PostPaymentProcessing too. That will cause duplicate processing.
+
+As we discussed, we need to keep the message handlers in the Sales endpoint to process in-flight messages: those sitting in the input queue and those sent before we apply the changes. But we must avoid getting new messages.
+
+To achieve that, we need to delete the Sales subscription to the SNS topic and modify the Sales endpoint configuration to stop subscribing to the `PaymentCompleted` event while keeping the message handler.
+
+Note that if you use NServiceBus, it will recreate the deleted subscription every time the endpoint restarts. To prevent that, the [`AutoSubscribe` feature comes with the option to disable subscriptions for specific events](https://docs.particular.net/nservicebus/messaging/publish-subscribe/controlling-what-is-subscribed?version=core_8#automatic-subscriptions-exclude-event-types-from-auto-subscribe). Thanks, [Tim](https://timbussmann.github.io/), for the reminder.
```
Comment (kbaley) [inline _posts/topology-changes.md:73]: Twitter might be a better link for Tim, if he has one

File: _posts/topology-changes.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+## Conclusion
+
+Evolving a system, it's not only a matter of updating contracts and interfaces or upgrading data structures. Changes, like the topology once, can break the system's backward compatibility or alter the system's behaviors. As with all things distributed systems, it's essential to start by analyzing the desired changes impact on message senders and publishers, recipients and subscribers, and the infrastructure.
```
Comment (kbaley) [inline _posts/topology-changes.md:77]: > Changes, like the topology once

This doesn't read correctly but not sure what it should be

**Review by kbaley — COMMENTED** [review]
Fair warning: I did this after a couple of Proseccos so double-check my comments

File: _posts/topology-changes.md (hunk 3 above)
Comment (mauroservienti) [reply _posts/topology-changes.md:73]: He has, but he also wants to be anonymous 😃

File: _posts/topology-changes.md (hunk 4 above)
Comment (mauroservienti) [reply _posts/topology-changes.md:77]:
```suggestion
Evolving a system, it's not only a matter of updating contracts and interfaces or upgrading data structures. Topology changes can break the system's backward compatibility or alter the system's behaviors.

As with all things distributed systems, it's essential to start by analyzing the desired changes impact on message senders and publishers, recipients and subscribers, and the infrastructure.
```

**Pull request merged**